File: src/uuid/uuid.js
var Uuid = {
  random: function() {
    var buf = new Uint8Array(16);
    window.crypto.getRandomValues(buf);

    // Set the leftmost bits to 0100 to set version 4
    buf[6] &= 0x0F;
    buf[6] |= 0x40;

    // Set the leftmost bits to 10 to set variant 1
    buf[8] &= 0x3F;
    buf[8] |= 0x80;

    var h = function(i) {
      var out = i.toString(16);
      if (out.length < 2) {
        out = "0".concat(out);
      }
      return out;
    }

    var out = "".concat(
      h(buf[0]), h(buf[1]), h(buf[2]), h(buf[3]), "-",
      h(buf[4]), h(buf[5]), "-",
      h(buf[6]), h(buf[7]), "-",
      h(buf[8]), h(buf[9]), "-",
      h(buf[10]), h(buf[11]), h(buf[12]),
      h(buf[13]), h(buf[14]), h(buf[15])
    );

    return out;
  }
}

File: src/uuid/uuid.c
#include <stdio.h>
#include <stdint.h>
#include <openssl/rand.h>

#include <urweb.h>

#include "uuid.h"

uw_Basis_string uw_Uuid_random(uw_context ctx) {
  uint8_t buf[16];
  { FILE *urand = fopen("/dev/urandom", "r");
    if (!urand) {
      uw_error(ctx, FATAL, "Failed to open /dev/urandom");
    }

    size_t read = fread(&buf, 1, 16, urand);
    if (!read || read < 16) {
      fclose(urand);
      uw_error(ctx, BOUNDED_RETRY, "Failed to read from /dev/urandom");
    }

    int ok = fclose(urand);
    if (ok != 0) {
      uw_error(ctx, BOUNDED_RETRY, "Failed to close /dev/urandom");
    }
  }

  // Set the leftmost bits to 0100 to set version 4
  buf[6] &= 0x0F;
  buf[6] |= 0x40;

  // Set the leftmost bits to 10 to set variant 1
  buf[8] &= 0x3F;
  buf[8] |= 0x80;

  char *out = uw_malloc(ctx, 37);
  sprintf(out,
    "%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x",
    buf[0], buf[1], buf[2],  buf[3],  buf[4],  buf[5],  buf[6],  buf[7],
    buf[8], buf[9], buf[10], buf[11], buf[12], buf[13], buf[14], buf[15]
  );

  return out;
}
